cmd/fts/cmd/tasks.go

```diff
@@ -0,0 +1,232 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	client "github.com/textileio/powergate/api/client"
+	userPb "github.com/textileio/powergate/api/gen/powergate/user/v1"
+)
+
+type Task struct {
+	name     string
+	path     string
+	bytes    int64
+	isDir    bool
+	isHidden bool
+}
+
+type TaskResult struct {
+	task    Task
+	cid     string
+	jobId   string
+	stage   string
+	state   *client.WatchStorageJobsEvent
+	records []*userPb.StorageDealRecord
+	err     error
+}
+
+type RunConfig struct {
+	token           string
+	ipfsrevproxy    string
+	serverAddress   string
+	concurrent      int64
+	minDealBytes    int64
+	maxStagingBytes int64
+	dryRun          bool
+}
+
+func Run(tasks []Task, config RunConfig) chan TaskResult {
+
+	sc := StagingConfig{
+		maxStagingBytes: config.maxStagingBytes,
+		minDealBytes:    config.minDealBytes,
+	}
+
+	// wg is used to force the application wait for all goroutines to finish before exiting.
+	wg := sync.WaitGroup{}
+	// ch is a buffered channel for processing tasks
+	ch := make(chan bool, config.concurrent)
```

Comment:
You can also do:
`make(chan struct{}, config.concurrent)`, and `ch < struct{}{}`.

A `struct{}` type has size 0 so consumes no memory.
Of course, just mentioning this as a curious note since you might see it in the willd... it doesn't make any difference in this program.

Reply:
sweet, gone